### PR TITLE
bpo-31648: Improve ElementPath

### PR DESCRIPTION
**Pull request by scoder**
https://bugs.python.org/issue31648

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Improvements to path predicates in ElementTree:
+* Allow whitespace around predicate parts, i.e. "[a = 'text']" instead of requiring the less readable "[a='text']".
+* Add support for text comparison of the current node, like "[.='text']".
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst]: Add "Patch by Stefan Behnel".

File: Lib/test/test_xml_etree.py [hunk 2]
```diff
@@ -2237,6 +2237,32 @@ def test_findall(self):
...
 
+        self.assertEqual(summarize_list(e.findall(".//section[tag='subtext']")),
+            ['section'])
+        self.assertEqual(summarize_list(e.findall(".//section[tag ='subtext']")),
+            ['section'])
+        self.assertEqual(summarize_list(e.findall(".//section[tag= 'subtext']")),
+            ['section'])
+        self.assertEqual(summarize_list(e.findall(".//section[tag = 'subtext']")),
+            ['section'])
+        self.assertEqual(summarize_list(e.findall(".//section[ tag = 'subtext' ]")),
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:2248]: Wouldn't be enough to test just this variant?

File: Lib/test/test_xml_etree.py [hunk 3]
```diff
@@ -2237,6 +2237,32 @@ def test_findall(self):
...
+            ['section'])
+
+        self.assertEqual(summarize_list(e.findall(".//tag[.='subtext']")),
+                         ['tag'])
+        self.assertEqual(summarize_list(e.findall(".//tag[. ='subtext']")),
+                         ['tag'])
+        self.assertEqual(summarize_list(e.findall('.//tag[.= "subtext"]')),
+                         ['tag'])
+        self.assertEqual(summarize_list(e.findall('.//tag[ . = "subtext" ]')),
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:2257]: Wouldn't be enough to test just this variant?

**Review by serhiy-storchaka — COMMENTED** [review]
Add a What's New entry. Document the new feature and add a "versionchanged" directive in the documentation.

File: Lib/xml/etree/ElementPath.py [hunk 4]
```diff
@@ -188,14 +190,21 @@ def select(context, result):
...
+                    for e in elem.findall(tag):
+                        if "".join(e.itertext()) == value:
+                            yield elem
+                            break
+        else:
+            def select(context, result):
+                for elem in result:
+                    if "".join(elem.itertext()) == value:
                         yield elem
                         break
```
Comment (serhiy-storchaka) [inline Lib/xml/etree/ElementPath.py]: I think there shouldn't be a `break`. Add a test for this.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Lib/test/test_xml_etree.py (hunk 2 above)
Comment (scoder) [reply Lib/test/test_xml_etree.py:2248]: Whitebox testing, yes. Blackbox testing, no. It's short enough, so I chose the latter. In parsers, it's easy to accidentally forget about special cases.

File: Lib/test/test_xml_etree.py (hunk 3 above)
Comment (scoder) [reply Lib/test/test_xml_etree.py:2257]: Same comment applies.

File: Lib/xml/etree/ElementPath.py (hunk 4 above)
Comment (scoder) [reply Lib/xml/etree/ElementPath.py]: This is inside of a predicate, so we only care about "found" or "not found". The break was there before.

File: Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst (hunk 1 above)
Comment (scoder) [reply Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst]: Done

File: Lib/xml/etree/ElementPath.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/xml/etree/ElementPath.py]: The break was there before for breaking an internal loop. `findall()` should find all elements, not the first one.

**Comment by scoder** [comment]
Documentation updated.

File: Lib/xml/etree/ElementPath.py (hunk 4 above)
Comment (scoder) [reply Lib/xml/etree/ElementPath.py]: To clarify: the break is correct.

Comment (scoder) [reply Lib/xml/etree/ElementPath.py]: Ah, you're right. :)

Comment (scoder) [reply Lib/xml/etree/ElementPath.py]: Fixed.

File: Doc/whatsnew/3.7.rst [hunk 5]
```diff
@@ -281,6 +281,13 @@ Function :func:`~uu.encode` now accepts an optional *backtick*
 keyword argument.  When it's true, zeros are represented by ``'`'``
 instead of spaces.  (Contributed by Xiang Zhang in :issue:`30103`.)
 
+xml.etree
+---------
+
+ElementPath predicates in the :meth:`find` methods can now compare text
+of the current node with ``[. = "text"]``, not only text in children.
+Predicates also allow adding spaces for better readability.
+
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: Add "(Contributed by ...)" as in other entries.

File: Lib/xml/etree/ElementPath.py [hunk 6]
```diff
@@ -157,6 +157,8 @@ def prepare_predicate(next, token):
             return
         if token[0] == "]":
             break
+        if token == ('', ''):
```
Comment (serhiy-storchaka) [inline Lib/xml/etree/ElementPath.py:160]: Is it for whitespaces? Please add a comment.

File: Lib/xml/etree/ElementPath.py [hunk 7]
```diff
@@ -188,16 +190,22 @@ def select(context, result):
...
+                    for e in elem.findall(tag):
+                        if "".join(e.itertext()) == value:
+                            yield elem
+                            break
+        else:
+            def select(context, result):
+                for elem in result:
+                    if "".join(elem.itertext()) == value:
                         yield elem
-                        break
```
Comment (serhiy-storchaka) [inline Lib/xml/etree/ElementPath.py:200]: Please add a special test case for this bug.

Comment (serhiy-storchaka) [reply Lib/xml/etree/ElementPath.py:200]: Ah, it is already added!

File: Doc/whatsnew/3.7.rst (hunk 5 above)
Comment (scoder) [reply Doc/whatsnew/3.7.rst]: Done.

File: Lib/xml/etree/ElementPath.py (hunk 6 above)
Comment (scoder) [reply Lib/xml/etree/ElementPath.py:160]: Done.

File: Lib/xml/etree/ElementPath.py (hunk 7 above)
Comment (scoder) [reply Lib/xml/etree/ElementPath.py:200]: Already done.

File: Doc/whatsnew/3.7.rst [hunk 8]
```diff
@@ -281,6 +281,14 @@ Function :func:`~uu.encode` now accepts an optional *backtick*
 keyword argument.  When it's true, zeros are represented by ``'`'``
 instead of spaces.  (Contributed by Xiang Zhang in :issue:`30103`.)
 
+xml.etree
+---------
+
+ElementPath predicates in the :meth:`find` methods can now compare text
+of the current node with ``[. = "text"]``, not only text in children.
+Predicates also allow adding spaces for better readability.
+(Contributed by Stefan Behnel.)
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: Add a reference to the issue.

File: Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst [hunk 9]
```diff
@@ -0,0 +1,4 @@
+Improvements to path predicates in ElementTree:
+* Allow whitespace around predicate parts, i.e. "[a = 'text']" instead of requiring the less readable "[a='text']".
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst]: These lines are merged together in the generated HTML. Add empty lines before and after the enumeration list.

**Comment by scoder** [comment]
Updated.

File: Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst [hunk 10]
```diff
@@ -1,4 +1,7 @@
 Improvements to path predicates in ElementTree:
+
 * Allow whitespace around predicate parts, i.e. "[a = 'text']" instead of requiring the less readable "[a='text']".
+
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst]: I think this one is redundant.

File: Doc/whatsnew/3.7.rst [hunk 11]
```diff
@@ -281,6 +281,14 @@ Function :func:`~uu.encode` now accepts an optional *backtick*
 keyword argument.  When it's true, zeros are represented by ``'`'``
 instead of spaces.  (Contributed by Xiang Zhang in :issue:`30103`.)
 
+xml.etree
+---------
+
+ElementPath predicates in the :meth:`find` methods can now compare text
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: Make `ElementPath` a link:
```
:class:`~xml.etree.ElementPath`
```
Add a reference to `elementtree-xpath`:
```
:ref:`some text <elementtree-xpath>`
```

Comment (scoder) [reply Doc/whatsnew/3.7.rst]: I added a reference.

File: Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst (hunk 10 above)
Comment (scoder) [reply Misc/NEWS.d/next/Library/2017-09-30-10-45-12.bpo-31648.Cai7ji.rst]: Removed.

**Comment by scoder** [comment]
Updated the news entries.
If you have further ideas about doc changes, I think it might be faster for you to edit the files, instead of explaining what changes you would like to have.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**